### PR TITLE
etcd-tester: add rate limits to stresser

### PR DESCRIPTION
**Pull request by gyuho**

### CODE REVIEW

**Comment by gyuho** [comment]
need to try this PR in 5 node cluster. Please do not merge yet.

**Comment by gyuho** [comment]
was hanging when canceling the stressers. Will fix.

**Comment by gyuho** [comment]
/cc @xiang90 @heyitsanthony  PTAL.

Currently this code is deployed at http://dash.etcd.io/dashboard/db/functional-tests-5-node (failpoint is disabled for now).

QPS flag 5,000's compaction size would be between 200K and 300K.

```
2016-07-07 21:24:37.371622 I | etcd-tester: [round#0] compacting 172954 entries (timeout 10s)
2016-07-07 21:24:37.371630 I | etcd-tester: [round#0] compacting storage (current revision 182954, compact revision 172954)
...
2016-07-07 21:24:39.721078 I | etcd-tester: [round#1 case#0] injecting failure "kill all members"
...
2016-07-07 21:25:55.062118 I | etcd-tester: [round#1 case#10] succeed!
2016-07-07 21:25:55.062132 I | etcd-tester: [round#1] compacting 196135 entries (timeout 13s)
2016-07-07 21:25:55.062145 I | etcd-tester: [round#1] compacting storage (current revision 379089, compact revision 369089)
...
```


QPS flag 10,000's compaction size would be between 300K and 400K.

```
2016-07-07 21:42:26.421645 I | etcd-tester: [round#2 case#0] injecting failure "kill all members"
...
2016-07-07 21:43:51.606524 I | etcd-tester: [round#2 case#10] succeed!
2016-07-07 21:43:51.606539 I | etcd-tester: [round#2] compacting 364940 entries (timeout 17s)
```

Thanks.

File: tools/functional-tester/etcd-tester/stresser.go [hunk 1]
```diff
@@ -77,68 +80,75 @@ func (s *stresser) Stress() error {
...
-
-						// default:
-						// errors from stresser.Cancel method:
-						// rpc error: code = 1 desc = context canceled (type grpc.rpcError)
-						// rpc error: code = 2 desc = grpc: the client connection is closing (type grpc.rpcError)
-					}
-					if shouldContinue {
-						continue
+
+			go func() {
```
Comment (heyitsanthony) [inline tools/functional-tester/etcd-tester/stresser.go]: split this out into another function? this nesting is too deep

File: tools/functional-tester/etcd-tester/stresser.go [hunk 2]
```diff
@@ -77,68 +80,75 @@ func (s *stresser) Stress() error {
 	s.conn = conn
 	s.cancel = cancel
 	s.wg = wg
+	s.donec = make(chan struct{})
```
Comment (heyitsanthony) [inline tools/functional-tester/etcd-tester/stresser.go]: `s.donec = ctx.Done()`?

**Comment by gyuho** [comment]
@heyitsanthony @xiang90 PTAL.

I added more useful logs that have been useful for us to debug test failures:
- print number of keys to restore in mvcc
- print the IP that the etcd-tester client is sending its request to
- stop stresser before compaction
- add configurable rate-limiter

This branch is deployed at http://dash.etcd.io/dashboard/db/functional-tests-5-node

**Comment by gyuho** [comment]
nvm... getting more failures... will fix tonight.

**Comment by xiang90** [comment]
@gyuho Please remove all logging stuff. This pull request should only add the rate limiter. one thing at a time.

**Comment by gyuho** [comment]
@xiang90 Sure, I will only keep the rate limiter stuff. This still needs more work on rate limit.

**Comment by gyuho** [comment]
@xiang90 PTAL. Deployed this with 5-node cluster. Works well http://dash.etcd.io/dashboard/db/functional-tests-5-node.

Thanks.

**Comment by xiang90** [comment]
@gyuho Wait more time? Try to get this merged tomorrow morning?

**Comment by gyuho** [comment]
Sure, let's keep it running till tomorrow. thx

**Comment by xiang90** [comment]
let's get this merged?

**Comment by gyuho** [comment]
yeah over 450 rounds with no failure.

thanks.

**Pull request merged**